### PR TITLE
Fixes for LHCbDIRAC v10r0

### PR DESCRIPTION
**Pull request by chrisburr**
This PR started as dd9d2cb to use the new `lb-dirac` command instead of `lb-run LHCbDIRAC` so it is possible to run pre-releases of DIRAC with Ganga. ([LBCORE-1420](https://its.cern.ch/jira/browse/)) You can try this out by using this branch and editing your gangarc to contain:

```
LHCbDiracVersion = v10r0-pre17
```

I then found a few bugs which I've fixed, see the commit messages for details.

For the deprecations, I've only fixed the functions I found were broken when running `echo "Hello world"`. It's probably worth somebody looking to see if there are any others decorated with `deprecated` on these pages:
* https://github.com/DIRACGrid/DIRAC/blob/rel-v6r22/Interfaces/API/Dirac.py
* https://github.com/DIRACGrid/DIRAC/blob/rel-v6r22/Interfaces/API/Job.py

We're hoping to deploy LHCbDIRAC v10r0 on Monday so the fixes for the deprecations are somewhat urgent even if the other commits wait longer.

This PR is still WIP as I've included a hack in b929cf6 to workaround that face the "stable" LbEnv symlink hasn't been updated yet. I'll revert that commit when it's no longer needed (probably tomorrow).

### CODE REVIEW

File: ganga/GangaLHCb/__init__.py [hunk 1]
```diff
@@ -136,7 +136,7 @@ def postBootstrapHook():
     configDirac.setSessionValue('noInputDataBannedSites', [])
     configDirac.setSessionValue('RequireDefaultSE', False)
     configDirac.setSessionValue('proxyInitCmd', 'lhcb-proxy-init')
-    configDirac.setSessionValue('proxyInfoCmd', 'lhcb-proxy-info')
+    configDirac.setSessionValue('proxyInfoCmd', 'dirac-proxy-info')
```
Comment (mesmith75) [inline ganga/GangaLHCb/__init__.py:139]: why this?

Comment (chrisburr) [reply ganga/GangaLHCb/__init__.py:139]: LHCbDIRAC doesn't provide a function called `lhcb-proxy-info` so 'lhcb-proxy-info' is actually coming from the normal LbEnv environment causing DIRAC to be set up twice, i.e. equivalent to `lb-run LHCbDIRAC/prod lb-run LHCbDIRAC/prod exec dirac-proxy-info`.

I noticed as the new `lb-dirac` command resets the environment before sourcing DIRAC so it fails with `lhcb-proxy-info: command not found`.

File: setup.py [hunk 2]
```diff
@@ -93,7 +93,7 @@ def run(self):
           'google-auth-oauthlib',
           'requests>=2.23.0'
       ],
-      extra_require={'profiler' : ['memory_profiler'], 'LHCb' : ['LbDevTools']},
+      extras_require={'profiler' : ['memory_profiler'], 'LHCb' : ['LbDevTools']},
```
Comment (egede) [inline setup.py:96]: Is this a completely unrelated bug fix in the dependencies?

File: ganga/GangaLHCb/__init__.py (hunk 1 above)
Comment (egede) [reply ganga/GangaLHCb/__init__.py:139]: Should `lhcb-proxy-init` be changed in a similar way?

**Review by egede — COMMENTED** [review]
We should remember that Ganga is used with Dirac outside LHCb as well. How old version of Dirac will `submitJob` etc. work with? We have to make sure not to cut off other users.

**Comment by egede** [comment]
> We're hoping to deploy LHCbDIRAC v10r0 on Monday so the fixes for the deprecations are somewhat urgent even if the other commits wait longer.

This seems unreasonable to me. The work here is not finished yet, there then has to be a new Ganga release and you depend on the all Ganga sessions will have changed to this new version before you change. Can you delay by maybe a week?

File: ganga/GangaLHCb/__init__.py (hunk 1 above)
Comment (chrisburr) [reply ganga/GangaLHCb/__init__.py:139]: Also, you could avoid sourcing LbEnv by depending on [`LbDiracWrappers`](https://pypi.org/project/LbDiracWrappers/) in the LHCb extra in `setup.py`.

File: setup.py (hunk 2 above)
Comment (chrisburr) [reply setup.py:96]: Yes it's just a bugfix for a typo. I noticed as doing `pip install ".[LHCb]"` would say "no extra called LHCb": https://setuptools.readthedocs.io/en/latest/setuptools.html#new-and-changed-setup-keywords

File: ganga/GangaLHCb/__init__.py (hunk 1 above)
Comment (chrisburr) [reply ganga/GangaLHCb/__init__.py:139]: > Should lhcb-proxy-init be changed in a similar way?

No, LHCbDIRAC provides a (very slightly) customised version of `dirac-proxy-init`: https://gitlab.cern.ch/lhcb-dirac/LHCbDIRAC/-/blob/master/LHCbDIRAC/Core/scripts/lhcb-proxy-init.py

**Comment by chrisburr** [comment]
> We should remember that Ganga is used with Dirac outside LHCb as well. How old version of Dirac will submitJob etc. work with? We have to make sure not to cut off other users.

The new name was introduced 5 years ago in (Vanilla) DIRAC v6r13 and a deprecation notice has been printed when it is used for the last two years (since v6r20p17).

File: setup.py (hunk 2 above)
Comment (egede) [reply setup.py:96]: Not sure how this has worked for the LHCb installations for the past few months ...

Comment (mesmith75) [reply setup.py:96]: We haven't used the `LbDev` stuff or the `memory_profiler` in any releases. The `LbDev` stuff was added for future use with the new application environment - I am not sure what the status of that package is. Currently it is a bit of a hack by source LbEnv from CVMFS every time we do something.

**Comment by mesmith75** [comment]
We don't see any of those warnings unfortunately - they are hidden away in the subprocesses. Maybe we should try and catch them and flag them for experts.

**Comment by chrisburr** [comment]
I guessed that was happening. The stable symlink of LbEnv has now been updated so this is ready from my perspective.

For releasing v10: We had a meeting this morning and we'd hope to deploy on Tuesday or Wednesday. Does that seem achievable from your side? For future releases, we're now be able to run tests with Ganga for pre-release versions of LHCbDIRAC so you should have more notice next time something changes.

**Comment by egede** [comment]
> For releasing v10: We had a meeting this morning and we'd hope to deploy on Tuesday or Wednesday. Does that seem achievable from your side? For future releases, we're now be able to run tests with Ganga for pre-release versions of LHCbDIRAC so you should have more notice next time something changes.

If the testing doesn't throw up any problems, it can be merged today. We can then try to get somebody to try out the DEV version on Friday and Monday and then hopefully make the release then. Seems just possible without taking too much of a risk.

**Comment by egede** [comment]
Hmm, the GangaLHCb tests are failing as there is a problem with the python3 environment. In the tests I see.

```
+ docker run --privileged --name GangaLHCbPR-1640-5 gangalhcbtest:PR-1640-5
LibCvmfs version 2.4, revision 25
created virtual environment CPython3.6.8.final.0-64 in 13933ms
  creator CPython3Posix(dest=/root/venv, clear=False, global=False)
  seeder FromAppData(download=False, pip=latest, setuptools=latest, wheel=latest, via=copy, app_data_dir=/root/.local/share/virtualenv/seed-app-data/v1.0.1)
  activators PythonActivator,FishActivator,XonshActivator,CShellActivator,PowerShellActivator,BashActivator
/root/ganga/ganga/GangaLHCb/test/testrunner: /root/venv/bin/pip: /root/venv/bin/python: bad interpreter: No such file or directory
/root/ganga/ganga/GangaLHCb/test/testrunner: /root/venv/bin/pip: /root/venv/bin/python: bad interpreter: No such file or directory
/root/ganga/ganga/GangaLHCb/test/testrunner: /root/venv/bin/pip: /root/venv/bin/python: bad interpreter: No such file or directory
```
I then investigate how the docker container was created. In a successful test from yesterday, I see
```
+ docker build -t gangalhcbtest:PR-1641-3 -f /var/lib/jenkins/workspace/Ganga_Devs_ganga_PR-1641/ganga/GangaLHCb/test/Dockerfile .
Sending build context to Docker daemon  58.91MB

Step 1/8 : FROM gitlab-registry.cern.ch/lhcb-core/lbdocker/centos7-build:v3
 ---> 5dd756659c73
Step 2/8 : LABEL maintainer "Ulrik Egede <ulrik.egede@monash.edu>"
 ---> Running in 9ee68c78ed9e
Removing intermediate container 9ee68c78ed9e
 ---> 68c92eac8d35
Step 3/8 : USER root
 ---> Running in e20ebeb36d0a
Removing intermediate container e20ebeb36d0a
 ---> ed8a1de74f7b
Step 4/8 : WORKDIR /root
 ---> Running in 973dacaae1ee
Removing intermediate container 973dacaae1ee
 ---> 5328464f7041
Step 5/8 : RUN yum install -y python-virtualenv python3
 ---> Running in ae3dd5381390
Loaded plugins: changelog, fastestmirror, kernel-module, ovl, protectbase,
              : tsflags, versionlock
Determining fastest mirrors
252 packages excluded due to repository protections
Resolving Dependencies
--> Running transaction check
---> Package python-virtualenv.noarch 0:15.1.0-4.el7_7 will be installed
---> Package python3.x86_64 0:3.6.8-10.el7 will be installed
--> Processing Dependency: python3-libs(x86-64) = 3.6.8-10.el7 for package: python3-3.6.8-10.el7.x86_64
--> Processing Dependency: python3-setuptools for package: python3-3.6.8-10.el7.x86_64
--> Processing Dependency: python3-pip for package: python3-3.6.8-10.el7.x86_64
--> Processing Dependency: libpython3.6m.so.1.0()(64bit) for package: python3-3.6.8-10.el7.x86_64
--> Running transaction check
---> Package python3-libs.x86_64 0:3.6.8-10.el7 will be installed
--> Processing Dependency: libtirpc.so.1()(64bit) for package: python3-libs-3.6.8-10.el7.x86_64
---> Package python3-pip.noarch 0:9.0.3-7.el7_7 will be installed
---> Package python3-setuptools.noarch 0:39.2.0-10.el7 will be installed
--> Running transaction check
---> Package libtirpc.x86_64 0:0.2.4-0.16.el7 will be installed
--> Finished Dependency Resolution
Beginning Kernel Module Plugin
Finished Kernel Module Plugin

Dependencies Resolved

================================================================================
 Package                  Arch         Version              Repository     Size
================================================================================
Installing:
 python-virtualenv        noarch       15.1.0-4.el7_7       updates       1.7 M
 python3                  x86_64       3.6.8-10.el7         base           69 k
Installing for dependencies:
 libtirpc                 x86_64       0.2.4-0.16.el7       base           89 k
 python3-libs             x86_64       3.6.8-10.el7         base          7.0 M
 python3-pip              noarch       9.0.3-7.el7_7        updates       1.8 M
 python3-setuptools       noarch       39.2.0-10.el7        base          629 k

Transaction Summary
================================================================================
Install  2 Packages (+4 Dependent packages)
```

while for this test is seems like python3 is already available when we build the docker container today.

```
+ docker build -t gangalhcbtest:PR-1640-5 -f /var/lib/jenkins/workspace/Ganga_Devs_ganga_PR-1640/ganga/GangaLHCb/test/Dockerfile .
Sending build context to Docker daemon     59MB

Step 1/8 : FROM gitlab-registry.cern.ch/lhcb-core/lbdocker/centos7-build:v3
 ---> 5dd756659c73
Step 2/8 : LABEL maintainer "Ulrik Egede <ulrik.egede@monash.edu>"
 ---> Using cache
 ---> 923e82d3b578
Step 3/8 : USER root
 ---> Using cache
 ---> 16b3c9f6ee36
Step 4/8 : WORKDIR /root
 ---> Using cache
 ---> c24ff0c2e7e3
Step 5/8 : RUN yum install -y python-virtualenv python3
 ---> Using cache
 ---> 6969aca07da6
Step 6/8 : COPY . ganga
 ---> 4c9c489c95a0
Step 7/8 : ENTRYPOINT /bin/bash
 ---> Running in c8c48151d9f6
Removing intermediate container c8c48151d9f6
 ---> 39e6905354a3
Step 8/8 : ENTRYPOINT . /etc/entrypoint.sh /root/ganga/ganga/GangaLHCb/test/testrunner
 ---> Running in 1cd34cdf770b
Removing intermediate container 1cd34cdf770b
 ---> 491bff2fc498
Successfully built 491bff2fc498
Successfully tagged gangalhcbtest:PR-1640-5
```

I guess the underlying centos7 docker container from CERN that the LHCb one relies on has changed! I imagine that the testrunner https://github.com/ganga-devs/ganga/blame/develop/ganga/GangaLHCb/test/testrunner has to be updated to fix this (it has a very odd `rm` command that I do not really understand). I will not have time to look at this before the weekend as I am in a meeting all day Friday.

**Comment by fstagni** [comment]
Hello,
just on this

> We should remember that Ganga is used with Dirac outside LHCb as well. How old version of Dirac will `submitJob` etc. work with? We have to make sure not to cut off other users.

The DIRAC versions that used the old APIs are not anymore maintained, so other Ganga users would definitely not be affected. At the same time now LHCb is switching to DIRAC v7r0 (with the API changes), and this version is out for several months now, so other Ganga users would definitely adopt it sooner than later.

**Review by egede — APPROVED** [review]

**Comment by egede** [comment]
With this merged, it will go into the DEV version of Ganga tomorrow night. We can then schedule a new release. To make things smooth, I would hope for a delay in deploying v10 until Thursday or Friday.

**Comment by chaen** [comment]
certainly not on Friday :D
